RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,798 @@
+
+On visitJoin(), the core logic is started by flattening the JoinNode tables into a list. We also require to flatten all the predicates including Join predicate and filter predicate into a single filter list. We separate all the output variables and assignments against the connector and keep for further processing. This flatten result will keep in a data structure called MultiJoinNode.
+
+```
+class MultiJoinNode
+{
+    private final CanonicalJoinNode node;
+    private final Assignments assignments;
+    private final boolean isConnectorJoin;
+}
+```
+
+```
+public class CanonicalJoinNode
+        extends PlanNode
+{
+    private final List<PlanNode> sources;
+    private final JoinType type;
+    private final Set<EquiJoinClause> criteria;
+    private final Set<RowExpression> filters;
+    private final List<VariableReferenceExpression> outputVariables;
+}
+```
+
+Sample logic for flatten JoinNode is as follows.
+
+```
+private void flattenNode(PlanNode resolved)
+{
+    if (resolved instanceof FilterNode) {
+        // We pull up all Filters to the top of the join graph, these will be pushed down again by predicate pushdown
+        // We do this in hope of surfacing any TableScan nodes that can be combined
+        FilterNode filterNode = (FilterNode) resolved;
+        filters.add(filterNode.getPredicate());
+        flattenNode(filterNode.getSource());
+        return;
+    }
+
+    if (!(resolved instanceof JoinNode)) {
+        //TODO optimize
+        if (resolved instanceof ProjectNode) {
+            // Certain ProjectNodes can be 'inlined' into the parent Jdbc TableScan, e.g a CAST expression
+            // We will do this here while flattening the JoinNode if possible
+
+            // For now, we log the fact that we saw a ProjectNode and if identity projection, will continue
+            logger.info("Found ProjectNode [%s] above TableScan/Filter node %s", resolved, resolved.getSources());
+
+            //Only identity projections can be handled.
+            if (AssignmentUtils.isIdentity(((ProjectNode) resolved).getAssignments())) {
+                flattenNode(((ProjectNode) resolved).getSource());
+                return;
+            }
+        }
+        sources.add(resolved);
+        return;
+    }
+
+    JoinNode joinNode = (JoinNode) resolved;
+    if (joinNode.getType() != INNER || !determinismEvaluator.isDeterministic(joinNode.getFilter().orElse(TRUE_CONSTANT))) {
+        sources.add(resolved);
+        return;
+    }
+
+    flattenNode(joinNode.getLeft());
+    flattenNode(joinNode.getRight());
+    joinNode.getCriteria().stream()
+            .map(criteria -> toRowExpression(criteria, functionResolution))
+            .forEach(joinCriteriaFilters::add);
+    joinNode.getFilter().ifPresent(joinCriteriaFilters::add);
+}
+``` 
+
+Here the sources contain all the tables (irrespective of connector) of that JoinNode, filters contain all the predicate of that JoinNode. Now we have completed the flattening of JoinNode.
```

Comment:
Correct it
filters contain all the predicate of sundered nodes and  joinCriteriaFilters contains predicate of each JoinNode

Reply:
corrected